src/plugins/vis_type_vega/public/data_model/search_api.ts

```diff
@@ -54,33 +58,72 @@ export class SearchAPI {
   ) {}
 
-  search(searchRequests: SearchRequest[]) {
+  async search(searchRequests: SearchRequest[]) {
     const { search } = this.dependencies.search;
     const requestResponders: any = {};
 
     return combineLatest(
-      searchRequests.map((request) => {
-        const requestId = request.name;
-        const params = getSearchParamsFromRequest(request, {
-          getConfig: this.dependencies.uiSettings.get.bind(this.dependencies.uiSettings),
-        });
-
-        if (this.inspectorAdapters) {
-          requestResponders[requestId] = this.inspectorAdapters.requests.start(requestId, request);
-          requestResponders[requestId].json(params.body);
-        }
-
-        return search({ params }, { abortSignal: this.abortSignal }).pipe(
-          tap((data) => this.inspectSearchResult(data, requestResponders[requestId])),
-          map((data) => ({
-            name: requestId,
-            rawResponse: data.rawResponse,
-          }))
-        );
-      })
+      await Promise.all(
+        searchRequests.map(async (request) => {
+          const requestId = request.name;
+          const dataSourceId = !!request.data_source_name
+            ? await this.findDataSourceIdbyName(request.data_source_name)
+            : undefined;
+
+          const params = getSearchParamsFromRequest(request, {
+            getConfig: this.dependencies.uiSettings.get.bind(this.dependencies.uiSettings),
+          });
+
+          if (this.inspectorAdapters) {
+            requestResponders[requestId] = this.inspectorAdapters.requests.start(
+              requestId,
+              request
+            );
+            requestResponders[requestId].json(params.body);
+          }
+
+          const searchApiParams =
+            dataSourceId && this.dependencies.dataSourceEnabled
+              ? { params, dataSourceId }
+              : { params };
+
+          return search(searchApiParams, { abortSignal: this.abortSignal }).pipe(
+            tap((data) => this.inspectSearchResult(data, requestResponders[requestId])),
+            map((data) => ({
+              name: requestId,
+              rawResponse: data.rawResponse,
+            }))
+          );
+        })
+      )
     );
   }
 
+  async findDataSourceIdbyName(dataSourceName: string) {
+    if (!this.dependencies.dataSourceEnabled) {
+      throw new Error('data_source_name cannot be used because data_source.enabled is false');
+    }
+    const dataSources = await this.dataSourceFindQuery(dataSourceName);
+
+    if (dataSources.total !== 1) {
+      throw new Error(
+        `Expected exactly 1 result for data_source_name ${dataSourceName} but got ${dataSources.total} results`
+      );
+    }
+
+    return dataSources.savedObjects.pop()?.id;
+  }
+
+  async dataSourceFindQuery(dataSourceName: string) {
+    return await this.dependencies.savedObjectsClient.find<DataSourceAttributes>({
+      type: 'data-source',
+      perPage: 10,
+      search: `"${dataSourceName}"`,
```

Comment:
do we need the double quotes?

Reply:
Yes, the quotes ensure that all the words in the `data_source_name` are in the results. For example, if `Data Source A` and `Data Source B` were data source names, running the `find` query for `Data Source A` would return both names. Logic was also added to ensure the data source name matches the title.